taskgroup/group.go

```diff
@@ -0,0 +1,90 @@
+package taskgroup
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+// ErrTerminated returned if Group was already terminated.
+var ErrTerminated = errors.New("taskgroup: terminated")
+
+// Option to modify Group.
+type Option func(g *Group)
+
+// WithContext passes parent context to use for the Group.
+func WithContext(ctx context.Context) Option {
+	return func(g *Group) {
+		g.ctx = ctx
+	}
+}
+
+// New returns instance of the Group.
+func New(opts ...Option) *Group {
+	g := &Group{
+		waitErr: make(chan struct{}),
+	}
+	for _, opt := range opts {
+		opt(g)
+	}
+	if g.ctx == nil {
+		g.ctx = context.Background()
+	}
+	g.ctx, g.cancel = context.WithCancel(g.ctx)
+	return g
+}
+
+// Group manages set of tasks.
+// Unlike errgroup.Group it is safe to call Go after Wait. If Group didn't terminate yet
+// the behavior will be the same as for errgroup.Group, if it did Go will exit immediatly.
+//
+// Zero value is not a valid Group. Must be initialized using New.
+type Group struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	mu      sync.Mutex
+	err     error
+	waitErr chan struct{}
+	wg      sync.WaitGroup
+}
+
+// Go spawns new goroutine that will run f, unless Group was already terminated.
+// In the latter case ErrTerminated error will be returned.
+func (g *Group) Go(f func(ctx context.Context) error) error {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	if g.err != nil {
+		return ErrTerminated
+	}
+	g.wg.Add(1)
+	go func() {
+		err := f(g.ctx)
+		g.wg.Done()
```

Comment:
should this be in the last line of go func, say line 72?

Reply:
i thought that it is correct to have here as well, because of the additional waitErr. but probably it will be more correct to execute Done as a very last thing. will change, thanks